plan/builders.go

```diff
@@ -75,10 +75,13 @@ type Builder interface {
 	ProjectRemap(input Rel, remap []int32, exprs ...expr.Expression) (*ProjectRel, error)
 	// Deprecated: Use AggregateColumns(...).Remap() instead.
 	AggregateColumnsRemap(input Rel, remap []int32, measures []AggRelMeasure, groupByCols ...int32) (*AggregateRel, error)
+	// Deprecated: Use Aggregate(...) instead.
 	AggregateColumns(input Rel, measures []AggRelMeasure, groupByCols ...int32) (*AggregateRel, error)
 	// Deprecated: Use AggregateExprs(...).Remap() instead.
 	AggregateExprsRemap(input Rel, remap []int32, measures []AggRelMeasure, groups ...[]expr.Expression) (*AggregateRel, error)
+	// Deprecated: Use Aggregate(...) instead.
 	AggregateExprs(input Rel, measures []AggRelMeasure, groups ...[]expr.Expression) (*AggregateRel, error)
```

Comment:
Optional:  Is a Grouping something we could provide a builder for instead of requiring the caller to construct a slice of expression slices?

Reply:
If I was using the interface I'd prefer to just call:

`func (b *builder) Aggregate(input Rel, measures []AggRelMeasure, groupingExpressions []expr.Expression) (*AggregateRel, error) {
`

Then Aggregate could either:

1.  simply assign ascending values to the grouping references for each expression 1:1 (not gaining any benefit of the identical expressions being able to be deduped and other optimization savings)
2.  assign values to the grouping references for each expression.  If an expression duplicates one seen previously then it isn't added to the expression list and a previous grouping reference value is used instead.

The notion of duplication for the second approach could mark equivalence incorrectly.  (For instance, perhaps marking a literal 1 as identical to another one even though it is of a different unit.). However I'm not sure it matters.

Reply:
Adding to your suggestion a bit. How about we do introduce more API on AggregateRel like below?

```
func (b *builder) Aggregate(input Rel, measures []AggRelMeasure, groupingExpressions []expr.Expression) (*AggregateRel, error)

func (ar *AggregateRel) addCube(groupingReferences []uint32) 
func (ar *AggregateRel) addRollup(groupingReferences []uint32)
func (ar *AggregateRel) addGroupingSet(groupingReferences []uint32)
```

Reply:
I do like the idea of adding a builder to ease construction.  I would try to avoid having to deal with grouping references unless the builder is going to tell me which reference a given expression is.  Perhaps we can return the references as part of adding the expressions?

```
func (arb *AggregateBuilder) addExpression(expr.Expression) uint32
```